File: scratch/PYTHON/xbox_remote_demo/controller_state.py
"""
This file contains the definition of ControllerState, an object which contains the current state of an Xbox controller:
"""


class ControllerState:
    """
    ControllerState represents the state of an entire Xbox controller at a given moment in time.
    """

    def __init(self):
        # A, B, X, and Y buttons. 0 is off and 1 is on.
        self.cn_a = 0
        self.cn_b = 0
        self.cn_x = 0
        self.cn_y = 0

        # Left and right bumpers. 0 is off and 1 is on.
        self.cn_left_bumper = 0
        self.cn_right_bumper = 0

        # Middle, back, and start buttons, whatever those are. 0 is off and 1 is on.
        self.cn_middle = 0
        self.cn_back = 0
        self.cn_start = 0

        # Left joystick x and y axes. Normalized between -1 and 1 as a float. 0 is neutral.
        self.cn_left_stick_x = 0.0
        self.cn_left_stick_y = 0.0

        # Right joystick x and y axes. Normalized between -1 and 1 as a float. 0 is neutral.
        self.cn_right_stick_x = 0.0
        self.cn_right_stick_y = 0.0

        # Left and right trigger, whatever those are. 0 is off and 1 is on.
        self.cn_left_trigger = 0
        self.cn_right_trigger = 0

        # Up, down, left, and right on the D-Pad. 0 is off and 1 is on.
        self.cn_dpad_up = 0
        self.cn_dpad_down = 0
        self.cn_dpad_left = 0
        self.cn_dpad_down = 0

File: development/JAVA_DEV/src/main/java/com/github/zeldazach/binghamtonrover/controller/ControllerState.java
package com.github.zeldazach.binghamtonrover.controller;

import java.util.Observable;

/**
 * ControllerState represents the state of the XBox controller at a given moment in time.
 */
public class ControllerState extends Observable {

    /**
     * The margin about a floating-point number in which other floats are considered to be equal.
     */
    public static final float EQUALITY_MARGIN = 1.0E-6f;

    /**
     * The threshold about 0.0 at which a stick movement will actually be processed.
     * This is fixed for now, but could be calibrated at runtime if necessary.
     */
    public static final float STICK_THRESHOLD = 0.02f;

    /**
     * Simple buttons, have an off state (false) and an on state (true).
     *
     * Select is the left-most middle button, mode is the glowing button with the logo, and start is the right-most middle button.
     * LThumb and RThumb are the left and right sticks when pressed.
     */
    public boolean buttonX, buttonY, buttonA, buttonB, buttonSelect, buttonMode, buttonStart, buttonLBumper, buttonRBumper, buttonLThumb, buttonRThumb;

    /**
     * Normalized values (between -1 and 1) for "continuous" axes.
     *
     * lTrigger and rTrigger are from -1 (not pressed) to 1 (fully pressed).
     *
     * The dpad values are interesting. They rage from 0 (meaning none pressed) to 1 (meaning left pressed).
     * Up is 0.25, right is 0.5, down is 0.75, and left is 1.0. If two adjacent are pressed together, the values is
     * between those listed above. Chart:
     *
     * Left + Up     : 0.125
     * Up            : 0.25
     * Up + Right    : 0.375
     * Right         : 0.5
     * Right + Down  : 0.675
     * Down          : 0.75
     * Down + Left   : 0.875
     * Left          : 1.0
     */
    public float lStickX, lStickY, rStickX, rStickY, lTrigger, rTrigger, dpad;

    public void update(String name, float value) {
        switch (name) {
            case "X":
                buttonX = float_to_bool(value);
                break;
            case "Y":
                buttonY = float_to_bool(value);
                break;
            case "A":
                buttonA = float_to_bool(value);
                break;
            case "B":
                buttonB = float_to_bool(value);
                break;
            case "Select":
                buttonSelect = float_to_bool(value);
                break;
            case "Mode":
                buttonMode = float_to_bool(value);
                break;
            case "Start":
                buttonStart = float_to_bool(value);
                break;
            case "Left Thumb":
                buttonLBumper = float_to_bool(value);
                break;
            case "Right Thumb":
                buttonRBumper = float_to_bool(value);
                break;
            case "Left Thumb 3":
                buttonLThumb = float_to_bool(value);
                break;
            case "Right Thumb 3":
                buttonRThumb = float_to_bool(value);
                break;
            case "x":
                lStickX = clamp_stick_value(value);
                break;
            case "y":
                lStickY = clamp_stick_value(value);
                break;
            case "rx":
                rStickX = clamp_stick_value(value);
                break;
            case "ry":
                rStickY = clamp_stick_value(value);
                break;
            case "z":
                lTrigger = value;
                break;
            case "rz":
                rTrigger = value;
                break;
            case "pov":
                dpad = value;
                break;
            default:
                throw new RuntimeException("ControllerState received update with unknown button name \"" + name + "\"");
        }

        setChanged();
        notifyObservers();
    }

    private boolean float_to_bool(float value) {
        if (Math.abs(value - 0.0f) < EQUALITY_MARGIN) {
            return false;
        }

        return true;
    }

    private float clamp_stick_value(float value) {
        if (Math.abs(value) < STICK_THRESHOLD) {
            return 0;
        }

        return value;
    }
}
